### PR TITLE
Improve performance of Job controller delete event handler

### PR DESCRIPTION
**Pull request by hakuna-matatah**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

 >Also, what this also tell us is basically there is still value in delete event handler sync path as well. :D
 
 Following up on this comment of mine ^^^ https://github.com/kubernetes/kubernetes/pull/127228#issuecomment-2343917121

This PR helps improve the Job controller performance in sync path for delete event handler. 
Job Update event handler performance was improved in https://github.com/kubernetes/kubernetes/pull/127228
With that we were able to get `50%` improvement in throughput.

With delete event handler, I'm able to see another `50%` improvement , thus we see overall `100%` improvement w.r.t event handler throughput 

See below for 10k+ jobs 

```
sh-4.2$ curl -ik https://localhost:10257/metrics | grep "pending_notifications"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP cache_pending_notifications [ALPHA] Current number of pending notifications
# TYPE cache_pending_notifications gauge
cache_pending_notifications{name="job-controller",resources="job"} 0
cache_pending_notifications{name="job-controller",resources="pod"} 2
cache_pending_notifications{name="ttl-after-finished-controller",resources="job"} 0
100  247k    0  247k    0     0  4127k      0 --:--:-- --:--:-- --:--:-- 4196k
sh-4.2$

```

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/pull/127228#issuecomment-2343917121

#### Special notes for your reviewer:

```
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="0.1"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="0.2"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="0.4"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="0.8"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="1.6"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="3.2"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="6.4"} 0
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="12.8"} 4371
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="25.6"} 51230
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="51.2"} 118775
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="102.4"} 126659
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="204.8"} 127744
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="409.6"} 127869
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="819.2"} 127930
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="1638.4"} 128036
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="3276.8"} 128087
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="6553.6"} 128112
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="13107.2"} 128127
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="26214.4"} 128130
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="52428.8"} 128130
job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete",le="+Inf"} 128132
job_controller_event_handling_duration_microseconds_sum{resource="jobs",type="delete"} 4.82433e+06
```
And p99 for processing delete event seems to be between 0.1024ms - 0.2048 ms as opposed to the data presented in tests from this [PR](https://github.com/kubernetes/kubernetes/pull/127228) is `26.2144 ms`   for `job_controller_event_handling_duration_microseconds_bucket{resource="jobs",type="delete` 

That shows an improvement of `99.2X` in performance and `50%` in throughput.

#### Does this PR introduce a user-facing change?

```release-note
Improve performance of the job controller when handling job delete events.
```

### CODE REVIEW

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -628,37 +631,60 @@ func (jm *Controller) syncOrphanPod(ctx context.Context, key string) error {
...
-		if job != nil {
-			// Skip cleanup of finalizers for pods owned by a job managed by an external controller
-			if controllerName := managedByExternalController(job); controllerName != nil {
-				logger.V(2).Info("Skip cleanup of the job finalizer for a pod owned by a job that is managed by an external controller", "key", key, "podUID", sharedPod.UID, "jobUID", job.UID, "controllerName", controllerName)
-				return nil
-			}
-		}
-		if job != nil && !util.IsJobFinished(job) {
-			// The pod was adopted. Do not remove finalizer.
+		if job != nil && (managedByExternalController(job) != nil || !util.IsJobFinished(job)) {
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:683]: Are these changes required? If not, please revert.

**Review by mimowo — COMMENTED** [review]
Initial pass to provide early feedback. I would be leaning to include the fix as it improves the performance significantly, but wondering still about the best approach.

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:683]: Basically cleaned up this section. Simplified it, not related to performance improvement.

**Comment by mimowo** [comment]
/cc @atiratree 
for another pair of eyes

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -628,37 +631,60 @@ func (jm *Controller) syncOrphanPod(ctx context.Context, key string) error {
...
-			logger.V(4).Info("Orphan pod has been deleted", "pod", key)
+			logger.V(4).Info("Orphan pod has been deleted", "pod", name)
 			return nil
 		}
 		return err
 	}
-	// Make sure the pod is still orphaned.
 	if controllerRef := metav1.GetControllerOf(sharedPod); controllerRef != nil {
 		if controllerRef.Kind != controllerKind.Kind || controllerRef.APIVersion != batch.SchemeGroupVersion.String() {
-			// The pod is controlled by an owner that is not a batch/v1 Job. Do not remove finalizer.
```
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:646]: As part of simplifying this function i ended up removing not realizing it :) 
https://github.com/kubernetes/kubernetes/pull/127378#discussion_r1761479201

I can add the comment back.

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -513,7 +512,11 @@ func (jm *Controller) deleteJob(logger klog.Logger, obj interface{}) {
 			return
 		}
 	}
-	jm.cleanupPodFinalizers(jobObj)
+	selector, err := metav1.LabelSelectorAsSelector(jobObj.Spec.Selector)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("unable to parse selector"))
```
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:517]: Cleaning up orphan pods through Job itself is a very rare case i.e; underlying assumption is that Pod delete event somehow drops, so we rely on Job delete event in such cases.
On top of that, failing parsing selector is something I can't think why it might happen given I couldn't think of a situation where we can't  parse selectors because  APIServer validation layer on creates/updates for a resource FMU.
If anything this will mask any issues in Validation layer IMO if we put a patch here instead of catching it ? wdyt ?

Also, Deployment controller returns error when it can't parse selectors for  [example](https://github.com/kubernetes/kubernetes/blob/v1.31.0/pkg/controller/deployment/deployment_controller.go#L532-L535))


That said, as a safe-guard we can call `jm.cleanupPodFinalizers(jobObj)`  if you think strongly about it.

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -628,37 +631,60 @@ func (jm *Controller) syncOrphanPod(ctx context.Context, key string) error {
 	}()
 
 	ns, name, err := cache.SplitMetaNamespaceKey(key)
+
+	if err == nil {
+		// If splitting is successful, treat the key as a pod name.
```
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:636]: Sorry, I don't fully understand your point ? 
Are you asking me to add a comment here in the code for readability ?
For key it will be for single resource i.e; `ns/name` and for selectors it will be of format "app=<app-name>"

File: pkg/controller/job/job_controller.go [hunk 5]
```diff
@@ -628,37 +631,60 @@ func (jm *Controller) syncOrphanPod(ctx context.Context, key string) error {
 	}()
 
 	ns, name, err := cache.SplitMetaNamespaceKey(key)
```
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:633]: >have another queue just for label selectors
 
 having another queue would make it clean but doesn't it mean another `x` number of go routines for this purpose because generally we depend on `concurrent-job-sync` ?
 Given its a rare scenario, I think it would be an overkill.
 
 >have some type prefix (key:, or selector:) and first parse the prefix
 
 
 
 We can do this if we want to be more explicit about the differentiation. In this if we want to pass the `ns` however for `selectors`, we might have to little bit more than just `key:, or selector:` ?  i.e for `key` it will be `key:ns/name` for `selector` it will be something like `selector:<ns>:<selector` . May be something similar you are also suggesting in option 3 above ? This way we can also use `ns` when looking up in pod store [here](https://github.com/kubernetes/kubernetes/pull/127378/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R654-R655) 
 
 
 
 Let me know, once we agree on this direction, i will make those specific changes.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:633]: A prefix sounds like enough.

Also note that a queue can be made of any other comparable type https://github.com/kubernetes/kubernetes/blob/3ab0c84bacd059a5b5995d72706c63b46ed24c0b/staging/src/k8s.io/client-go/util/workqueue/rate_limiting_queue.go#L93

so you could have something like this:

```
type orphanPodKey struct {
  kind string
  value string
}
```

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/job/job_controller.go:704]: Please revert it then. If you have an idea for a cleanup then feel free to open a dedicated PR, but I think we should keep the comment `logger.V(2).Info("Skip cleanup` which might be relevant in some scenarios.

File: pkg/controller/job/job_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/job/job_controller.go]: Please do

File: pkg/controller/job/job_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/job/job_controller.go]: Yeah, my comment was to gain confidence if there is no overlap in the format of the key, and the label selector. The label selector I think has a bit richer syntax than app=<app-name>, like `'environment in (production),tier in (frontend)`, but indeed the formats looks disjoint as they all require spaces. Still, relying on the formats being disjoint seems fragile, and we have a couple of ideas in the [thread](https://github.com/kubernetes/kubernetes/pull/127378#discussion_r1761422311) how to not rely on this. Let's continue the discussion there.

File: pkg/controller/job/job_controller_test.go [hunk 6]
```diff
@@ -6756,19 +6765,54 @@ func TestSyncOrphanPod(t *testing.T) {
...
+				// Ensure cache is synced before calling syncOrphanPod
+				if !cache.WaitForCacheSync(ctx.Done(), podInformer.HasSynced) {
+					t.Fatalf("Failed to wait for pod informer cache sync")
+				}
+				orphanPodKey := orphanPodKey{
+					kind:  "selector",
+					value: fmt.Sprintf("%s:%s", orphanPod.Namespace, selector),
+				}
+				// Sync orphan pod by selector
+				err = manager.syncOrphanPod(ctx, orphanPodKey)
```
Comment (mimowo) [inline pkg/controller/job/job_controller_test.go:6794]: nit: Move the `syncOrphanPod` invocation out of the branches

File: pkg/controller/job/job_controller.go [hunk 7]
```diff
@@ -615,27 +628,68 @@ func (jm *Controller) processNextOrphanPod(ctx context.Context) bool {
...
 	}
+	for _, pod := range pods {
+		if err := jm.handleSingleOrphanPod(ctx, pod.Namespace, pod.Name); err != nil {
+			logger.Error(err, "syncing orphan pod failed", "pod", pod.Name)
+		}
+	}
+	return nil
+}
 
+// handleSingleOrphanPod processes a single orphan pod (logic remains unchanged).
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:686]:
```suggestion
// handleSingleOrphanPod processes a single orphan pod
```

File: pkg/controller/job/job_controller.go [hunk 8]
```diff
@@ -143,6 +144,13 @@ type syncJobCtx struct {
 	ready                           int32
 }
 
+type orphanPodKey struct {
+	// Either "key" or "selector"
+	kind string
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:149]: I don't feel strongly about it, but I would prefer to have enum consts for this. I think then we can get support for static code analysis for the `switch` statement in code.

File: pkg/controller/job/job_controller_test.go [hunk 9]
```diff
@@ -6756,19 +6765,54 @@ func TestSyncOrphanPod(t *testing.T) {
...
-			err = manager.syncOrphanPod(ctx, cache.MetaObjectToName(orphanPod).String())
-			if err != nil {
-				t.Fatalf("Failed sync orphan pod: %v", err)
+				// Simulate syncing using pod selector
+				selector, err := metav1.LabelSelectorAsSelector(tc.podSelector)
+				if err != nil {
+					t.Fatalf("Error parsing pod label selector: %v", err)
+				}
+				// Ensure cache is synced before calling syncOrphanPod
+				if !cache.WaitForCacheSync(ctx.Done(), podInformer.HasSynced) {
```
Comment (mimowo) [inline pkg/controller/job/job_controller_test.go:6786]: Is it needed? It is a bit surprising why it is only on the selector path. If needed I would suggest to explain in the comment.

**Review by mimowo — COMMENTED** [review]
LGTM overall, left a couple of nits.

**Comment by mimowo** [comment]
Actually, with this new mechanism we could 'revert' the flow changes in the previous PR for handling Job updates. 

IIUC the previous PR changed the flow so that in some scenarios we start deleting finalizers only when the Job object itself is deleted. While I'm not aware of any specific bug due to this, I would be leaning to revert the flow as it is already reasonably well tested.

In that case we would just queue orphan pod cleanup by selector from update job, if finished.

 WDYT @alculquicondor @atiratree @hakuna-matatah ?

File: pkg/controller/job/job_controller_test.go (hunk 9 above)
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller_test.go:6786]: More of a safety check before proceeding further. Its not a must and can be removed for consitency.

Comment (mimowo) [reply pkg/controller/job/job_controller_test.go:6786]: Then I would suggest to remove, unless you observed some flakes before adding it.

File: pkg/controller/job/job_controller.go [hunk 10]
```diff
@@ -143,6 +144,13 @@ type syncJobCtx struct {
...
+	// Either "namespace/name" (for kind key) or "namespace:selector" (for kind selector)
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go]: if there is a namespace for both, I suggest storing it in its own field, rather than concatenating it.

File: pkg/controller/job/job_controller.go [hunk 11]
```diff
@@ -513,7 +521,12 @@ func (jm *Controller) deleteJob(logger klog.Logger, obj interface{}) {
 			return
 		}
 	}
-	jm.cleanupPodFinalizers(jobObj)
+	selector, err := metav1.LabelSelectorAsSelector(jobObj.Spec.Selector)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("job %s/%s has invalid label selector: %w", jobObj.Namespace, jobObj.Name, err))
+
+	}
+	jm.enqueueLabelSelector(jobObj.Namespace, selector.String())
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:529]: I would just inline it here.

File: pkg/controller/job/job_controller.go [hunk 12]
```diff
@@ -568,7 +581,7 @@ func (jm *Controller) enqueueOrphanPod(obj *v1.Pod) {
 		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
 		return
 	}
-	jm.orphanQueue.Add(key)
+	jm.enqueuePodKey(key)
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:584]: just inline

File: pkg/controller/job/job_controller.go [hunk 13]
```diff
@@ -620,22 +633,63 @@ func (jm *Controller) processNextOrphanPod(ctx context.Context) bool {
...
+	}
+
+	// Fetch all pods that match the label selector.
+	// relatively expensive operation but it is called only from the orphan reconciler
+	pods, err := jm.podStore.Pods(namespace).List(selector)
 	if err != nil {
 		return err
 	}
+	for _, pod := range pods {
+		if err := jm.handleSingleOrphanPod(ctx, pod.Namespace, pod.Name); err != nil {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:679]: why do another get if you already have the object?
```suggestion
		if err := jm.handleSingleOrphanPod(ctx, pod); err != nil {
```

Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:688]: Not sure i understand your comment, If we don't do it it here and just send Pod object, we might have to do it [here](https://github.com/kubernetes/kubernetes/pull/127378/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R698) again  iiuc. Am i missing anything ?

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:688]: why? Just don't do a get again.

Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:688]: I think for logging [here](https://github.com/kubernetes/kubernetes/pull/127378/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R716) we are accessing name and ns.

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:688]: which is already accessible via `pod.Name` and `pod.Namespace`

File: pkg/controller/job/job_controller.go [hunk 14]
```diff
@@ -143,6 +144,22 @@ type syncJobCtx struct {
...
+	OrphanPodKeyKindNamespaceName orphanPodKeyKind = iota
+	// "selector"
+	OrphanPodKeyKindSelector
+)
+
+type orphanPodKey struct {
+	// Either "key" or "selector"
+	kind orphanPodKeyKind
+	// Either "namespace/name" or "namespace:selector"
+	value string
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:160]: Please split namespace into its own field, so that we can avoid concatenation.

File: pkg/controller/job/job_controller.go (hunk 13 above)
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:688]: Oh, i initially misunderstood what you meant, I think you are referring to this [get](https://github.com/kubernetes/kubernetes/pull/127378/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R698) operation which can be avoided. If so, that makes sense to me.

Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go:688]: It also looks like there is one more [place](https://github.com/kubernetes/kubernetes/pull/127378/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R658) where this is being called and it looks like it doesn't have Pod object there.

File: pkg/controller/job/job_controller.go (hunk 10 above)
Comment (hakuna-matatah) [reply pkg/controller/job/job_controller.go]: We don't technically need the namespace for key kind: name in orphan reconciler operations. However, during the PR discussion, we organically agreed to include it for logging purposes since we're already supplying it for key kind: selector. Given that, it made sense to also include it for key kind: name, so moving the namespace to its own field seems right to me now.

File: pkg/controller/job/job_controller.go (hunk 13 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:683]: You can refactor the part of `handleSingleOrphanPod` that already deals with a `pod` into another function.

**Comment by hakuna-matatah** [comment]
> Actually, with this new mechanism we could 'revert' the flow changes in the previous PR for handling Job updates. 
> 
> IIUC the previous PR changed the flow so that in some scenarios we start deleting finalizers only when the Job object itself is deleted. While I'm not aware of any specific bug due to this, I would be leaning to revert the flow as it is already reasonably well tested.
> 
> In that case we would just queue orphan pod cleanup by selector from update job, if finished.
> 
> WDYT @alculquicondor @atiratree @hakuna-matatah ? (could be a follow up)

I am not sure what flow changes you are referring specifically when you are talking about update event handler.
FMU we haven't changed anything functionally. Instead of synchronously removing the finalizers for pod(S) that is associated to Job which is either in finished state or pod that has no owner(orphan), we just moved it to reconciler to improve performance in sync path and for additional safeguard purposes. 

I don't really see a strong need to add  to orphan reconciler explicitly from correctness point of view for update event handler unless I am missing something here.
It is not like we are enqueuing to the queue explicitly like we do in delete event handler, we are just picking up on what is already found in the queue and removing finalizers for finished Job and returning it instead of going further into the reconciler code path. 
By adding to OrphanQueue explicitly in updateJob handler it doesn't guarantee that it won't be part of regular queue. So we would end up seeing the same item twice in two different queues though we might only act on one ?

**Comment by mimowo** [comment]
@hakuna-matatah thank you for the summary.

> I don't really see a strong need to add to orphan reconciler explicitly from correctness point of view for update event handler unless I am missing something here. It is not like we are enqueuing to the queue explicitly like we do in delete event handler, we are just picking up on what is already found in the queue and removing finalizers for finished Job and returning it instead of going further into the reconciler code path. By adding to OrphanQueue explicitly in updateJob handler it doesn't guarantee that it won't be part of regular queue. So we would end up seeing the same item twice in two different queues though we might only act on one ?

I agree - I also don't see a correctness issue with how it is done after that PR.

> However from consistency stand point b/w update and delete handler it makes sense to me. I don't have a strong opinion on this tbh.

My intention was to motivate it by consistency with the previous approach, but the consistency between deleteJob and updateJob is a better argument. 

If we have an agreement I would still suggest as a follow up PR for cleanup. WDYT @alculquicondor ?

File: pkg/controller/job/job_controller_test.go [hunk 15]
```diff
@@ -6756,19 +6765,49 @@ func TestSyncOrphanPod(t *testing.T) {
...
+					t.Fatalf("Creating orphan pod: %v", err)
+				}
+				orphanKey = orphanPodKey{
+					kind:      OrphanPodKeyKindName,
+					namespace: orphanPod.Namespace,
+					value:     orphanPod.Name,
+				}
 			}
-			err = manager.syncOrphanPod(ctx, cache.MetaObjectToName(orphanPod).String())
+			// Sync orphan pod by selector
```
Comment (mimowo) [inline pkg/controller/job/job_controller_test.go:6807]:
```suggestion
			// Sync orphan pod by name or selector
```
nit

File: pkg/controller/job/job_controller_test.go [hunk 16]
```diff
@@ -6756,19 +6765,49 @@ func TestSyncOrphanPod(t *testing.T) {
...
+			err = manager.syncOrphanPod(ctx, orphanKey)
 			if err != nil {
-				t.Fatalf("Failed sync orphan pod: %v", err)
+				t.Fatalf("Failed to sync orphan pod by label selector: %v", err)
```
Comment (mimowo) [inline pkg/controller/job/job_controller_test.go:6810]:
```suggestion
				t.Fatalf("Failed to sync orphan pod: %v", err)
```
nit, since it can be either name or selector.

**Comment by alculquicondor** [comment]
> By adding to OrphanQueue explicitly in updateJob handler it doesn't guarantee that it won't be part of regular queue. So we would end up seeing the same item twice in two different queues though we might only act on one ?

There is still a minor benefit: cleaning up finalizers is a lower priority than starting new jobs. So by keeping finalizer removal out of the main queue, we could theoretically be serving new jobs faster.
We could also "optimize" updateJob handler to not add the job to the main queue if it's already marked as finished, as we know there is nothing to do.

File: pkg/controller/job/job_controller.go (hunk 7 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:698]: Please address this.

File: pkg/controller/job/job_controller_test.go [hunk 17]
```diff
@@ -6756,19 +6765,49 @@ func TestSyncOrphanPod(t *testing.T) {
...
+				orphanPod, err = clientset.CoreV1().Pods("default").Create(ctx, orphanPod, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("Creating orphan pod: %v", err)
+				}
+				orphanKey = orphanPodKey{
+					kind:      OrphanPodKeyKindName,
+					namespace: orphanPod.Namespace,
+					value:     orphanPod.Name,
+				}
 			}
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:6806]: You can get less repeated code this way

```golang
podBuilder := buildPod().name(name).deletionTimestamp().trackingFinalizer()
if tc.owner != nil {
	podBuilder = podBuilder.owner(*tc.owner)
}
orphanKey := orphanPodKey{
	kind:      OrphanPodKeyKindName,
	namespace: orphanPod.Namespace,
	value:     orphanPod.Name,
}
if tc.podSelector != nil {
	podBuilder = podBuilder.labels(tc.podSelector.MatchLabels)
	selector, err := metav1.LabelSelectorAsSelector(tc.podSelector)
	if err != nil {
		t.Fatalf("Error parsing pod label selector: %v", err)
	}
	orphanKey = orphanPodKey{
		kind:      OrphanPodKeyKindSelector,
		namespace: orphanPod.Namespace,
		value:     selector.String(),
	}
}
orphanPod = podBuilder.Pod
orphanPod, err = clientset.CoreV1().Pods("default").Create(ctx, orphanPod, metav1.CreateOptions{})
if err != nil {
	t.Fatalf("Creating orphan pod: %v", err)
}
```

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
Thanks!

**Comment by mimowo** [comment]
Right, so we have two small things in favor: consistency between the handlers, and off-loading the main queue. I'm for it, @hakuna-matatah if you want to follow up with the PR I can review it.
